<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>RNA-seq Analysis Example</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">RNA-seq Analysis Example</h1>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This vignette demonstrates how to use <code>thoth</code> for a
typical RNA-seq analysis workflow using edgeR and limma. We’ll:</p>
<ol style="list-style-type: decimal">
<li>Set up a reproducible project structure</li>
<li>Track raw count data with DVC</li>
<li>Perform quality control</li>
<li>Run differential expression analysis</li>
<li>Track decisions and results</li>
<li>Create a reproducible pipeline</li>
</ol>
</div>
<div id="project-setup" class="section level2">
<h2>Project Setup</h2>
<p>First, let’s create a new project with DVC enabled:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(thoth)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(edgeR)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">library</span>(limma)</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">library</span>(statmod)  <span class="co"># Required for edgeR</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="co"># Create new project</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="fu">create_analytics_project</span>(</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>  <span class="st">&quot;rnaseq_analysis&quot;</span>,</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>  <span class="at">use_dvc =</span> <span class="cn">TRUE</span>,</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>  <span class="at">use_docker =</span> <span class="cn">TRUE</span>,</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>  <span class="at">git_init =</span> <span class="cn">TRUE</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>)</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="co"># Change to project directory</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">&quot;rnaseq_analysis&quot;</span>)</span></code></pre></div>
</div>
<div id="load-and-track-raw-data" class="section level2">
<h2>Load and Track Raw Data</h2>
<p>Let’s start with a simulated RNA-seq dataset:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Simulate count data</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>n_genes <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>n_samples <span class="ot">&lt;-</span> <span class="dv">8</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co"># Create counts matrix</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>  <span class="fu">rnbinom</span>(n_genes <span class="sc">*</span> n_samples, <span class="at">mu =</span> <span class="dv">100</span>, <span class="at">size =</span> <span class="dv">1</span>),</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>  <span class="at">nrow =</span> n_genes,</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>  <span class="at">ncol =</span> n_samples</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>)</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co"># Add row and column names</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="fu">rownames</span>(counts) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;gene_&quot;</span>, <span class="dv">1</span><span class="sc">:</span>n_genes)</span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="fu">colnames</span>(counts) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;sample_&quot;</span>, <span class="dv">1</span><span class="sc">:</span>n_samples)</span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co"># Create sample metadata</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a>sample_info <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a>  <span class="at">sample =</span> <span class="fu">colnames</span>(counts),</span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a>  <span class="at">group =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;control&quot;</span>, <span class="st">&quot;treatment&quot;</span>), <span class="at">each =</span> <span class="dv">4</span>),</span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a>  <span class="at">batch =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;B&quot;</span>), <span class="dv">4</span>)</span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a>)</span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a></span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a><span class="co"># Save and track raw data with DVC</span></span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a>counts <span class="sc">|&gt;</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">&quot;gene_id&quot;</span>) <span class="sc">|&gt;</span></span>
<span id="cb2-28"><a href="#cb2-28" tabindex="-1"></a>  <span class="fu">write_csv_dvc</span>(</span>
<span id="cb2-29"><a href="#cb2-29" tabindex="-1"></a>    <span class="st">&quot;data/raw/counts.csv&quot;</span>,</span>
<span id="cb2-30"><a href="#cb2-30" tabindex="-1"></a>    <span class="at">message =</span> <span class="st">&quot;Add raw count data&quot;</span></span>
<span id="cb2-31"><a href="#cb2-31" tabindex="-1"></a>  )</span>
<span id="cb2-32"><a href="#cb2-32" tabindex="-1"></a></span>
<span id="cb2-33"><a href="#cb2-33" tabindex="-1"></a>sample_info <span class="sc">|&gt;</span></span>
<span id="cb2-34"><a href="#cb2-34" tabindex="-1"></a>  <span class="fu">write_csv_dvc</span>(</span>
<span id="cb2-35"><a href="#cb2-35" tabindex="-1"></a>    <span class="st">&quot;data/raw/sample_info.csv&quot;</span>,</span>
<span id="cb2-36"><a href="#cb2-36" tabindex="-1"></a>    <span class="at">message =</span> <span class="st">&quot;Add sample metadata&quot;</span></span>
<span id="cb2-37"><a href="#cb2-37" tabindex="-1"></a>  )</span></code></pre></div>
</div>
<div id="initialize-decision-tree" class="section level2">
<h2>Initialize Decision Tree</h2>
<p>Let’s track our analytical decisions:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Initialize decision tracking</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>decision_file <span class="ot">&lt;-</span> <span class="fu">initialize_decision_tree</span>(</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="at">analysis_id =</span> <span class="st">&quot;rnaseq_2024&quot;</span>,</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  <span class="at">analyst =</span> <span class="st">&quot;Data Scientist&quot;</span>,</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>  <span class="at">description =</span> <span class="st">&quot;Differential expression analysis of treatment vs control&quot;</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="data-processing-and-qc" class="section level2">
<h2>Data Processing and QC</h2>
<div id="create-dgelist-object" class="section level3">
<h3>1. Create DGEList Object</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Read data</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>counts_df <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;data/raw/counts.csv&quot;</span>)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>sample_info <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;data/raw/sample_info.csv&quot;</span>)</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co"># Create DGEList</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>dge <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>  <span class="at">counts =</span> counts_df <span class="sc">|&gt;</span> <span class="fu">column_to_rownames</span>(<span class="st">&quot;gene_id&quot;</span>),</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>  <span class="at">group =</span> sample_info<span class="sc">$</span>group</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>)</span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co"># Add batch information</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>dge<span class="sc">$</span>samples<span class="sc">$</span>batch <span class="ot">&lt;-</span> sample_info<span class="sc">$</span>batch</span></code></pre></div>
</div>
<div id="quality-control" class="section level3">
<h3>2. Quality Control</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Calculate library sizes and CPM</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>lib_sizes <span class="ot">&lt;-</span> dge<span class="sc">$</span>samples<span class="sc">$</span>lib.size</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>cpms <span class="ot">&lt;-</span> <span class="fu">cpm</span>(dge)</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co"># Filter low expression genes</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">filterByExpr</span>(dge, <span class="at">group =</span> dge<span class="sc">$</span>samples<span class="sc">$</span>group)</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>dge_filtered <span class="ot">&lt;-</span> dge[keep, ]</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co"># Record filtering decision</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="fu">record_decision</span>(</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a>  decision_file,</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a>  <span class="at">check =</span> <span class="st">&quot;Gene filtering&quot;</span>,</span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>  <span class="at">observation =</span> <span class="fu">sprintf</span>(</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a>    <span class="st">&quot;Removed %d low expression genes&quot;</span>,</span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a>    <span class="fu">sum</span>(<span class="sc">!</span>keep)</span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a>  ),</span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a>  <span class="at">decision =</span> <span class="st">&quot;Filter using filterByExpr()&quot;</span>,</span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a>  <span class="at">reasoning =</span> <span class="st">&quot;Remove genes with consistently low counts&quot;</span>,</span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a>  <span class="at">evidence =</span> <span class="st">&quot;plots/gene_expression_density.png&quot;</span></span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a>)</span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a></span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a><span class="co"># Save filtered data</span></span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a>dge_filtered<span class="sc">$</span>counts <span class="sc">|&gt;</span></span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">&quot;gene_id&quot;</span>) <span class="sc">|&gt;</span></span>
<span id="cb5-26"><a href="#cb5-26" tabindex="-1"></a>  <span class="fu">write_csv_dvc</span>(</span>
<span id="cb5-27"><a href="#cb5-27" tabindex="-1"></a>    <span class="st">&quot;data/processed/filtered_counts.csv&quot;</span>,</span>
<span id="cb5-28"><a href="#cb5-28" tabindex="-1"></a>    <span class="at">message =</span> <span class="st">&quot;Add filtered count data&quot;</span>,</span>
<span id="cb5-29"><a href="#cb5-29" tabindex="-1"></a>    <span class="at">stage_name =</span> <span class="st">&quot;filter_genes&quot;</span>,</span>
<span id="cb5-30"><a href="#cb5-30" tabindex="-1"></a>    <span class="at">deps =</span> <span class="fu">c</span>(</span>
<span id="cb5-31"><a href="#cb5-31" tabindex="-1"></a>      <span class="st">&quot;data/raw/counts.csv&quot;</span>,</span>
<span id="cb5-32"><a href="#cb5-32" tabindex="-1"></a>      <span class="st">&quot;data/raw/sample_info.csv&quot;</span></span>
<span id="cb5-33"><a href="#cb5-33" tabindex="-1"></a>    ),</span>
<span id="cb5-34"><a href="#cb5-34" tabindex="-1"></a>    <span class="at">params =</span> <span class="fu">list</span>(</span>
<span id="cb5-35"><a href="#cb5-35" tabindex="-1"></a>      <span class="at">min_cpm =</span> <span class="dv">1</span>,</span>
<span id="cb5-36"><a href="#cb5-36" tabindex="-1"></a>      <span class="at">min_samples =</span> <span class="dv">4</span></span>
<span id="cb5-37"><a href="#cb5-37" tabindex="-1"></a>    )</span>
<span id="cb5-38"><a href="#cb5-38" tabindex="-1"></a>  )</span>
<span id="cb5-39"><a href="#cb5-39" tabindex="-1"></a></span>
<span id="cb5-40"><a href="#cb5-40" tabindex="-1"></a><span class="co"># Normalization</span></span>
<span id="cb5-41"><a href="#cb5-41" tabindex="-1"></a>dge_filtered <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(dge_filtered)</span>
<span id="cb5-42"><a href="#cb5-42" tabindex="-1"></a></span>
<span id="cb5-43"><a href="#cb5-43" tabindex="-1"></a><span class="co"># Record normalization decision</span></span>
<span id="cb5-44"><a href="#cb5-44" tabindex="-1"></a><span class="fu">record_decision</span>(</span>
<span id="cb5-45"><a href="#cb5-45" tabindex="-1"></a>  decision_file,</span>
<span id="cb5-46"><a href="#cb5-46" tabindex="-1"></a>  <span class="at">check =</span> <span class="st">&quot;Normalization&quot;</span>,</span>
<span id="cb5-47"><a href="#cb5-47" tabindex="-1"></a>  <span class="at">observation =</span> <span class="st">&quot;Library sizes vary between samples&quot;</span>,</span>
<span id="cb5-48"><a href="#cb5-48" tabindex="-1"></a>  <span class="at">decision =</span> <span class="st">&quot;Apply TMM normalization&quot;</span>,</span>
<span id="cb5-49"><a href="#cb5-49" tabindex="-1"></a>  <span class="at">reasoning =</span> <span class="st">&quot;Account for composition bias&quot;</span>,</span>
<span id="cb5-50"><a href="#cb5-50" tabindex="-1"></a>  <span class="at">evidence =</span> <span class="cn">NULL</span></span>
<span id="cb5-51"><a href="#cb5-51" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="quality-control-plots" class="section level3">
<h3>3. Quality Control Plots</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># MDS plot</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>limma<span class="sc">::</span><span class="fu">plotMDS</span>(dge_filtered) <span class="sc">|&gt;</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ggsave</span>(</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>    <span class="st">&quot;plots/mds_plot.png&quot;</span>,</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>    <span class="at">width =</span> <span class="dv">8</span>,</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>    <span class="at">height =</span> <span class="dv">6</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>  )</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co"># Record MDS observation</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="fu">record_decision</span>(</span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>  decision_file,</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>  <span class="at">check =</span> <span class="st">&quot;Sample clustering&quot;</span>,</span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a>  <span class="at">observation =</span> <span class="st">&quot;Samples cluster by treatment with some batch effect&quot;</span>,</span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a>  <span class="at">decision =</span> <span class="st">&quot;Include batch in design matrix&quot;</span>,</span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a>  <span class="at">reasoning =</span> <span class="st">&quot;Account for technical variation&quot;</span>,</span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a>  <span class="at">evidence =</span> <span class="st">&quot;plots/mds_plot.png&quot;</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
<div id="differential-expression-analysis" class="section level2">
<h2>Differential Expression Analysis</h2>
<div id="design-matrix-and-model-fitting" class="section level3">
<h3>1. Design Matrix and Model Fitting</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># Create design matrix</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>  <span class="sc">~</span>batch <span class="sc">+</span> group,</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>  <span class="at">data =</span> dge_filtered<span class="sc">$</span>samples</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>)</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="co"># Fit model</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(</span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>  <span class="fu">voom</span>(dge_filtered, design),</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a>  design</span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a>)</span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">eBayes</span>(fit)</span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a><span class="co"># Get results</span></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">topTable</span>(</span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a>  fit,</span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a>  <span class="at">coef =</span> <span class="st">&quot;grouptreatment&quot;</span>,</span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a>  <span class="at">number =</span> <span class="cn">Inf</span></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a>) <span class="sc">|&gt;</span></span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">&quot;gene_id&quot;</span>)</span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a></span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a><span class="co"># Save results with DVC</span></span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a>results <span class="sc">|&gt;</span></span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a>  <span class="fu">write_csv_dvc</span>(</span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a>    <span class="st">&quot;data/processed/de_results.csv&quot;</span>,</span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a>    <span class="at">message =</span> <span class="st">&quot;Add differential expression results&quot;</span>,</span>
<span id="cb7-27"><a href="#cb7-27" tabindex="-1"></a>    <span class="at">stage_name =</span> <span class="st">&quot;de_analysis&quot;</span>,</span>
<span id="cb7-28"><a href="#cb7-28" tabindex="-1"></a>    <span class="at">deps =</span> <span class="st">&quot;data/processed/filtered_counts.csv&quot;</span>,</span>
<span id="cb7-29"><a href="#cb7-29" tabindex="-1"></a>    <span class="at">params =</span> <span class="fu">list</span>(</span>
<span id="cb7-30"><a href="#cb7-30" tabindex="-1"></a>      <span class="at">adj_p_threshold =</span> <span class="fl">0.05</span>,</span>
<span id="cb7-31"><a href="#cb7-31" tabindex="-1"></a>      <span class="at">lfc_threshold =</span> <span class="dv">1</span></span>
<span id="cb7-32"><a href="#cb7-32" tabindex="-1"></a>    ),</span>
<span id="cb7-33"><a href="#cb7-33" tabindex="-1"></a>    <span class="at">metrics =</span> <span class="cn">TRUE</span>  <span class="co"># Track as DVC metrics</span></span>
<span id="cb7-34"><a href="#cb7-34" tabindex="-1"></a>  )</span>
<span id="cb7-35"><a href="#cb7-35" tabindex="-1"></a></span>
<span id="cb7-36"><a href="#cb7-36" tabindex="-1"></a><span class="co"># Record analysis decisions</span></span>
<span id="cb7-37"><a href="#cb7-37" tabindex="-1"></a><span class="fu">record_decision</span>(</span>
<span id="cb7-38"><a href="#cb7-38" tabindex="-1"></a>  decision_file,</span>
<span id="cb7-39"><a href="#cb7-39" tabindex="-1"></a>  <span class="at">check =</span> <span class="st">&quot;Differential expression&quot;</span>,</span>
<span id="cb7-40"><a href="#cb7-40" tabindex="-1"></a>  <span class="at">observation =</span> <span class="fu">sprintf</span>(</span>
<span id="cb7-41"><a href="#cb7-41" tabindex="-1"></a>    <span class="st">&quot;Found %d DE genes (FDR &lt; 0.05, |logFC| &gt; 1)&quot;</span>,</span>
<span id="cb7-42"><a href="#cb7-42" tabindex="-1"></a>    <span class="fu">sum</span>(results<span class="sc">$</span>adj.P.Val <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> <span class="fu">abs</span>(results<span class="sc">$</span>logFC) <span class="sc">&gt;</span> <span class="dv">1</span>)</span>
<span id="cb7-43"><a href="#cb7-43" tabindex="-1"></a>  ),</span>
<span id="cb7-44"><a href="#cb7-44" tabindex="-1"></a>  <span class="at">decision =</span> <span class="st">&quot;Use voom-limma pipeline&quot;</span>,</span>
<span id="cb7-45"><a href="#cb7-45" tabindex="-1"></a>  <span class="at">reasoning =</span> <span class="st">&quot;Account for mean-variance relationship&quot;</span>,</span>
<span id="cb7-46"><a href="#cb7-46" tabindex="-1"></a>  <span class="at">evidence =</span> <span class="st">&quot;data/processed/de_results.csv&quot;</span></span>
<span id="cb7-47"><a href="#cb7-47" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="visualization" class="section level3">
<h3>2. Visualization</h3>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Create volcano plot</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>results <span class="sc">|&gt;</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> logFC, <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(adj.P.Val))) <span class="sc">+</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>  <span class="fu">geom_point</span>(</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">color =</span> adj.P.Val <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> <span class="fu">abs</span>(logFC) <span class="sc">&gt;</span> <span class="dv">1</span>),</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>    <span class="at">alpha =</span> <span class="fl">0.6</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;red&quot;</span>)) <span class="sc">+</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">&quot;Volcano Plot&quot;</span>,</span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;log2 Fold Change&quot;</span>,</span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;-log10 Adjusted P-value&quot;</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a>  <span class="fu">ggsave</span>(</span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a>    <span class="st">&quot;plots/volcano_plot.png&quot;</span>,</span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a>    <span class="at">width =</span> <span class="dv">10</span>,</span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a>    <span class="at">height =</span> <span class="dv">8</span></span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a>  )</span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a></span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a><span class="co"># Create heatmap of top DE genes</span></span>
<span id="cb8-22"><a href="#cb8-22" tabindex="-1"></a>top_genes <span class="ot">&lt;-</span> results <span class="sc">|&gt;</span></span>
<span id="cb8-23"><a href="#cb8-23" tabindex="-1"></a>  <span class="fu">filter</span>(adj.P.Val <span class="sc">&lt;</span> <span class="fl">0.05</span>, <span class="fu">abs</span>(logFC) <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<span id="cb8-24"><a href="#cb8-24" tabindex="-1"></a>  <span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">50</span>) <span class="sc">|&gt;</span></span>
<span id="cb8-25"><a href="#cb8-25" tabindex="-1"></a>  <span class="fu">pull</span>(gene_id)</span>
<span id="cb8-26"><a href="#cb8-26" tabindex="-1"></a></span>
<span id="cb8-27"><a href="#cb8-27" tabindex="-1"></a><span class="co"># Get normalized expression for top genes</span></span>
<span id="cb8-28"><a href="#cb8-28" tabindex="-1"></a>expr_mat <span class="ot">&lt;-</span> <span class="fu">voom</span>(dge_filtered, design)<span class="sc">$</span>E[top_genes, ]</span>
<span id="cb8-29"><a href="#cb8-29" tabindex="-1"></a></span>
<span id="cb8-30"><a href="#cb8-30" tabindex="-1"></a><span class="co"># Save heatmap</span></span>
<span id="cb8-31"><a href="#cb8-31" tabindex="-1"></a><span class="fu">png</span>(<span class="st">&quot;plots/heatmap.png&quot;</span>, <span class="at">width =</span> <span class="dv">800</span>, <span class="at">height =</span> <span class="dv">1000</span>)</span>
<span id="cb8-32"><a href="#cb8-32" tabindex="-1"></a><span class="fu">Heatmap</span>(</span>
<span id="cb8-33"><a href="#cb8-33" tabindex="-1"></a>  expr_mat,</span>
<span id="cb8-34"><a href="#cb8-34" tabindex="-1"></a>  <span class="at">name =</span> <span class="st">&quot;Expression&quot;</span>,</span>
<span id="cb8-35"><a href="#cb8-35" tabindex="-1"></a>  <span class="at">column_split =</span> dge_filtered<span class="sc">$</span>samples<span class="sc">$</span>group,</span>
<span id="cb8-36"><a href="#cb8-36" tabindex="-1"></a>  <span class="at">show_row_names =</span> <span class="cn">FALSE</span></span>
<span id="cb8-37"><a href="#cb8-37" tabindex="-1"></a>)</span>
<span id="cb8-38"><a href="#cb8-38" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
</div>
</div>
<div id="export-analysis-documentation" class="section level2">
<h2>Export Analysis Documentation</h2>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># Export decision tree</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">export_decision_tree</span>(</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  decision_file,</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  <span class="at">format =</span> <span class="st">&quot;html&quot;</span>,</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>  <span class="at">output_path =</span> <span class="st">&quot;reports/analysis_decisions.html&quot;</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="complete-dvc-pipeline" class="section level2">
<h2>Complete DVC Pipeline</h2>
<p>Our analysis is now tracked with DVC:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># View pipeline</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="ex">dvc</span> dag</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="co"># Reproduce analysis</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="ex">dvc</span> repro</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="co"># Push to remote</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="ex">dvc</span> push</span></code></pre></div>
</div>
<div id="conclusion" class="section level2">
<h2>Conclusion</h2>
<p>This example demonstrates how to: - Track RNA-seq data with DVC -
Document QC and analysis decisions - Create reproducible bioinformatics
pipelines - Generate publication-ready figures - Share results and
methods</p>
<p>The entire analysis is version controlled and reproducible, with: -
Raw data tracked by DVC - Code in Git - Dependencies managed by renv -
Analysis decisions documented - Docker environment for
reproducibility</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
