<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Decision Tracking with thoth</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Decision Tracking with thoth</h1>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>One of the challenges in bioinformatics and data analysis is
documenting and tracking human decisions made during the analysis
process. While code and data can be version controlled, the reasoning
behind key analytical decisions often remains undocumented. The
<code>thoth</code> package provides functionality to track these
decisions in a structured way.</p>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>To install the package from GitHub, first ensure you have all the
required dependencies:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># Install required dependencies if not already installed</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>required_packages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;usethis&quot;</span>, <span class="st">&quot;yaml&quot;</span>, <span class="st">&quot;cli&quot;</span>, <span class="st">&quot;digest&quot;</span>, <span class="st">&quot;rmarkdown&quot;</span>, <span class="st">&quot;tools&quot;</span>)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="cf">for</span> (pkg <span class="cf">in</span> required_packages) {</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(pkg, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>    <span class="fu">install.packages</span>(pkg)</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>  }</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>}</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="co"># Install thoth from GitHub</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;devtools&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>}</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;sebrauschert/thoth&quot;</span>)</span></code></pre></div>
</div>
<div id="getting-started" class="section level2">
<h2>Getting Started</h2>
<p>First, let’s load the package:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">library</span>(thoth)</span></code></pre></div>
</div>
<div id="creating-a-decision-tree" class="section level2">
<h2>Creating a Decision Tree</h2>
<p>When starting a new analysis, initialize a decision tree to track
your decisions:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>decision_file <span class="ot">&lt;-</span> <span class="fu">initialize_decision_tree</span>(</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>  <span class="at">analysis_id =</span> <span class="st">&quot;RNA_seq_2024_01&quot;</span>,</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="at">analyst =</span> <span class="st">&quot;Jane Smith&quot;</span>,</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  <span class="at">description =</span> <span class="st">&quot;Differential expression analysis of treatment vs control samples&quot;</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="recording-decisions" class="section level2">
<h2>Recording Decisions</h2>
<p>As you progress through your analysis, record key decisions and
observations:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Record a quality control decision</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">record_decision</span>(</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  <span class="at">file_path =</span> decision_file,</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="at">check =</span> <span class="st">&quot;Sample-wise PCA clustering&quot;</span>,</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="at">observation =</span> <span class="st">&quot;Treatment samples cluster together except for sample T3&quot;</span>,</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  <span class="at">decision =</span> <span class="st">&quot;Exclude sample T3&quot;</span>,</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>  <span class="at">reasoning =</span> <span class="st">&quot;T3 clusters with controls, likely sample swap&quot;</span>,</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>  <span class="at">evidence =</span> <span class="st">&quot;plots/PCA_pre_filtering.pdf&quot;</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>)</span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co"># Record a biological validation step</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="fu">record_decision</span>(</span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a>  <span class="at">file_path =</span> decision_file,</span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a>  <span class="at">check =</span> <span class="st">&quot;Known pathway markers&quot;</span>,</span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a>  <span class="at">observation =</span> <span class="st">&quot;Expected stress response genes upregulated&quot;</span>,</span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a>  <span class="at">decision =</span> <span class="st">&quot;Results biologically plausible&quot;</span>,</span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a>  <span class="at">reasoning =</span> <span class="st">&quot;Key marker genes show expected direction of change&quot;</span>,</span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a>  <span class="at">evidence =</span> <span class="st">&quot;tables/marker_genes_expression.csv&quot;</span></span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="generating-methods-documentation" class="section level2">
<h2>Generating Methods Documentation</h2>
<p>You can automatically generate a methods section from your decision
tree:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Generate methods section</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>methods_text <span class="ot">&lt;-</span> <span class="fu">generate_methods_section</span>(decision_file)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="fu">cat</span>(methods_text)</span></code></pre></div>
</div>
<div id="exporting-decision-trees" class="section level2">
<h2>Exporting Decision Trees</h2>
<p>Export your decision tree to various formats for sharing or
documentation:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Export to markdown</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">export_decision_tree</span>(decision_file, <span class="at">format =</span> <span class="st">&quot;md&quot;</span>)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co"># Export to HTML (requires rmarkdown)</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">requireNamespace</span>(<span class="st">&quot;rmarkdown&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  <span class="fu">export_decision_tree</span>(decision_file, <span class="at">format =</span> <span class="st">&quot;html&quot;</span>)</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="integration-with-project-workflow" class="section level2">
<h2>Integration with Project Workflow</h2>
<p>The decision tracking functionality integrates seamlessly with your
analysis workflow:</p>
<ol style="list-style-type: decimal">
<li>Initialize a decision tree at the start of your analysis</li>
<li>Record decisions as you make them</li>
<li>Use the exported documentation for methods sections in papers</li>
<li>Keep the decision tree under version control</li>
</ol>
</div>
<div id="best-practices" class="section level2">
<h2>Best Practices</h2>
<p>Here are some recommended practices for using the decision tracking
functionality:</p>
<ol style="list-style-type: decimal">
<li><strong>Be Consistent</strong>: Record decisions as you make them,
not after the fact</li>
<li><strong>Be Specific</strong>: Include clear observations and
reasoning</li>
<li><strong>Link Evidence</strong>: Always link to supporting plots,
tables, or other evidence</li>
<li><strong>Version Control</strong>: Keep decision trees under version
control</li>
<li><strong>Review</strong>: Regularly review decisions with team
members</li>
</ol>
</div>
<div id="conclusion" class="section level2">
<h2>Conclusion</h2>
<p>By tracking analytical decisions in a structured way, we can:</p>
<ul>
<li>Make the analysis process more transparent</li>
<li>Facilitate knowledge transfer within teams</li>
<li>Improve reproducibility</li>
<li>Generate better documentation</li>
</ul>
<p>The decision tracking functionality in <code>thoth</code> helps make
human expertise as trackable as code, breaking free from the “90%
automation trap” where crucial human judgment remains hidden.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
